spring-cloud-gcp-data-datastore/src/main/java/org/springframework/cloud/gcp/data/datastore/core/DatastoreOperations.java

```diff
@@ -269,14 +272,35 @@
 	 */
 	DatastoreEntityConverter getDatastoreEntityConverter();
 
+	/**
+	 * Finds keys by using a Cloud Datastore query.
+	 * Resulting Slice can be used to get a Pageable for the next page or to determine if the next page exists.
+	 * @param query the query to execute.
+	 * @param entityClass the type of object to retrieve.
+	 * @param pageable that indicates page number and page size
+	 * @param <T> the type of object to retrieve.
+	 * @return a Slice containing found objects
+	 */
+	<T> Slice<Key> queryKeysSlice(KeyQuery query, Class<T> entityClass, Pageable pageable);
+
+	/**
+	 * Finds entities by using a Cloud Datastore query.
+	 * Resulting Slice can be used to get a Pageable for the next page or to determine if the next page exists.
+	 * @param query the query to execute.
+	 * @param entityClass the type of object to retrieve.
+	 * @param pageable that indicates page number and page size
+	 * @param <T> the type of object to retrieve.
+	 * @return a Slice containing found objects
+	 */
+	<T> Slice<T> queryEntitiesSlice(StructuredQuery query, Class<T> entityClass, Pageable pageable);
+
 	/**
 	 * Finds objects by using a Cloud Datastore query. If the query is a key-query, then keys are
 	 * returned.
 	 * @param query the query to execute.
 	 * @param entityClass the type of object to retrieve.
 	 * @param <T> the type of object to retrieve.
-	 * @return a list of the objects found. If no keys could be found the list will be
-	 * empty.
+	 * @return an iterable containing found objects and a cursor
 	 */
 	<T> DatastoreResultsIterable<?> queryKeysOrEntities(Query query, Class<T> entityClass);
```

Comment:
Can we also deprecate this method and bifurcate it like you did for `Slice`? This can be in a separate PR.

Reply:
created an issue: #2559